src/main/java/org/springframework/data/couchbase/repository/query/ReactiveStringBasedCouchbaseQuery.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.couchbase.repository.query;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.couchbase.core.CouchbaseOperations;
+import org.springframework.data.couchbase.core.ReactiveCouchbaseOperations;
+import org.springframework.data.couchbase.core.query.Query;
+import org.springframework.data.repository.core.NamedQueries;
+import org.springframework.data.repository.query.ParametersParameterAccessor;
+import org.springframework.data.repository.query.QueryMethodEvaluationContextProvider;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.util.Assert;
+
+/**
+ * Query to use a plain JSON String to create the {@link Query} to actually execute.
+ *
+ * @author Michael Reiche
+ * @since 4.1
+ */
+public class ReactiveStringBasedCouchbaseQuery extends AbstractReactiveCouchbaseQuery {
```

Comment:
this class looks very similar to the non-reactive one (same comments apply). Could some of the methods be refactored into a parent that extends the AbstractReactiveCouchbaseQuery (so a common string based one that has the common logic)

Reply:
The two classes are identical, other than one extends AbstractCouchbaseQuery and the other extends AbstractReactiveCouchbaseQuery - and the first two args of the constructor are the regular and reactive versions.  However, I don't see a clean way to combine the common methods while keeping the inherited classes different.

Reply:
Maybe interfaces with default method implementations?